domain/application/state/containerimagemetadata_test.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	coreresourcetesting "github.com/juju/juju/core/resources/testing"
+	"github.com/juju/juju/domain/application"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	schematesting "github.com/juju/juju/domain/schema/testing"
+	loggertesting "github.com/juju/juju/internal/logger/testing"
+)
+
+type containerImageMetadataSuite struct {
+	schematesting.ModelSuite
+}
+
+var _ = gc.Suite(&containerImageMetadataSuite{})
+
+func (s *containerImageMetadataSuite) TestContainerImageMetadataPutAndGet(c *gc.C) {
+	st := NewContainerImageMetadataState(s.TxnRunnerFactory(), loggertesting.WrapCheckLog(c))
+	resourceUUID := coreresourcetesting.GenResourceID(c)
+	ociImageMetadata := application.ContainerImageMetadata{
+		RegistryPath: "testing@sha256:beef-deed",
+		Username:     "docker-registry",
+		Password:     "fragglerock",
+	}
+	metadataUUID, err := st.Put(
+		context.Background(),
+		resourceUUID,
+		ociImageMetadata.RegistryPath,
+		ociImageMetadata.Username,
+		ociImageMetadata.Password,
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(metadataUUID, gc.Not(gc.Equals), "")
+
+	retrievedMetadata, err := st.Get(context.Background(), resourceUUID)
```

Comment:
**suggestion:** For me it is a bad practice to do a put and a get in a same Unit Test. It break the Unit test principle.

* Test put should verify that the db has been updated (and run sql against db in assertion)
* Test get should verify that we get expected data from defined db state (and we should arrange db before act)